indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java

```diff
@@ -0,0 +1,685 @@
+  {
+    return enabled;
+  }
+
+  private void scheduleQueuePoll(long delay)
+  {
+    executor.schedule(this::processBatchesDue, delay, TimeUnit.MILLISECONDS);
+  }
+
+  /**
+   * Queues a SegmentAllocateRequest. The returned future may complete successfully
+   * with a non-null value or with a non-null value.
+   */
+  public Future<SegmentIdWithShardSpec> add(SegmentAllocateRequest request)
+  {
+    if (!leaderSelector.isLeader()) {
+      throw new ISE("Cannot allocate segment if not leader.");
+    } else if (!isEnabled()) {
+      throw new ISE("Batched segment allocation is disabled.");
+    }
+
+    final AllocateRequestKey requestKey = new AllocateRequestKey(request, false);
+    final AtomicReference<Future<SegmentIdWithShardSpec>> requestFuture = new AtomicReference<>();
+
+    keyToBatch.compute(requestKey, (key, existingBatch) -> {
+      AllocateRequestBatch computedBatch = existingBatch;
+      if (computedBatch == null) {
+        computedBatch = new AllocateRequestBatch(key);
+        computedBatch.resetQueueTime();
+        processingQueue.offer(computedBatch);
+      }
+
+      // Possible race condition:
+      // t1 -> new batch is added to queue or batch already exists in queue
+      // t2 -> executor pops batch, processes all requests in it
+      // t1 -> new request is added to dangling batch and is never picked up
+      // Solution: For existing batch, call keyToBatch.remove() on the key to
+      // wait on keyToBatch.compute() to finish before proceeding with processBatch().
+      // For new batch, keyToBatch.remove() would not wait as key is not in map yet
+      // but a new batch is unlikely to be due immediately, so it won't get popped right away.
+      requestFuture.set(computedBatch.add(request));
+      return computedBatch;
+    });
+
+    return requestFuture.get();
+  }
+
+  private void requeueBatch(AllocateRequestBatch batch)
+  {
+    log.info("Requeueing [%d] failed requests in batch [%s].", batch.size(), batch.key);
+    keyToBatch.compute(batch.key, (key, existingBatch) -> {
+      if (existingBatch == null) {
+        batch.resetQueueTime();
+        processingQueue.offer(batch);
+        return batch;
+      } else {
+        // Merge requests from this batch to existing one
+        existingBatch.merge(batch);
+        return existingBatch;
+      }
+    });
+  }
+
+  private void processBatchesDue()
+  {
+    // If not leader, clear the queue and do not schedule any more rounds of processing
+    if (!leaderSelector.isLeader()) {
+      log.info("Not leader anymore. Clearing [%d] batches from queue.", processingQueue.size());
+      processingQueue.clear();
+      keyToBatch.clear();
+      return;
+    }
+
+    // Process all batches which are due
+    log.debug("Processing all batches which are due for execution.");
+    int numProcessedBatches = 0;
+
+    AllocateRequestBatch nextBatch = processingQueue.peek();
+    while (nextBatch != null && nextBatch.isDue()) {
```

Comment:
what if the head of the queue is not due but others in queue are?

Reply:
That cannot happen, the batch due time is nothing but queuedAtTime + maxWaitTime.
So the batch at the head of the queue has to be the one that will be processed first.
When requeueing a batch, it's queue time is reset.